### PR TITLE
[SPEC] Too restrictive data type

### PR DESCRIPTION
**Pull request by mbencer**

### CODE REVIEW

**Comment by mbencer** [comment]
According to spec `ShapeOf`: "Match (needs post-conversion of data_type: port[0]: i64->i32)"
@slyalin @GlebKazantaev @ilyachur What do you mean by it?

File: src/ngraph/op/broadcast.cpp [hunk 1]
```diff
@@ -115,10 +115,9 @@ void op::v1::Broadcast::validate_and_infer_types()
     // shape node should have integer data type. For now we only allow i64
     auto shape_et = get_input_element_type(1);
     NODE_VALIDATION_CHECK(this,
-                          shape_et.compatible(element::Type_t::i64),
-                          "Broadcast shape must have element type i64, but has ",
+                          shape_et.is_integral_number(),
+                          "Broadcast shape must have integral number element type, but has ",
```
Comment (tsocha) [inline src/ngraph/op/broadcast.cpp]:
```suggestion
                          "Broadcast shape must be an integral number, but is: ",
```

File: src/ngraph/op/broadcast.cpp [hunk 2]
```diff
@@ -131,10 +130,9 @@ void op::v1::Broadcast::validate_and_infer_types()
         // axes_mapping node should have integer data type. For now we only allow i64
         auto axes_et = get_input_element_type(2);
         NODE_VALIDATION_CHECK(this,
-                              axes_et.compatible(element::Type_t::i64),
-                              "Broadcast axes must have element type i64, but has ",
+                              axes_et.is_integral_number(),
+                              "Broadcast axes must have integral number element type, but has ",
```
Comment (tsocha) [inline src/ngraph/op/broadcast.cpp]:
```suggestion
                              "Broadcast axes must be integral numbers, but are: ",
```

File: src/ngraph/op/experimental/layers/interpolate.cpp [hunk 3]
```diff
@@ -35,8 +35,8 @@ op::Interpolate::Interpolate(const Output<Node>& image,
 void op::Interpolate::validate_and_infer_types()
 {
     NODE_VALIDATION_CHECK(this,
-                          get_input_element_type(1).compatible(element::Type_t::i64),
-                          "output shape must have element type i64.");
+                          get_input_element_type(1).is_integral_number(),
+                          "output shape must have integral number element type.");
```
Comment (tsocha) [inline src/ngraph/op/experimental/layers/interpolate.cpp]:
```suggestion
                          "output shape must be an integral number.");
```

Comment (tsocha) [reply src/ngraph/op/experimental/layers/interpolate.cpp]: Should we print a provided type as well as above?

File: src/ngraph/op/experimental/layers/prior_box.cpp [hunk 4]
```diff
@@ -37,14 +37,14 @@ void op::PriorBox::validate_and_infer_types()
     // shape node should have integer data type. For now we only allow i64
     auto layer_shape_et = get_input_element_type(0);
     NODE_VALIDATION_CHECK(this,
-                          layer_shape_et.compatible(element::Type_t::i64),
-                          "layer shape input must have element type i64, but has ",
+                          layer_shape_et.is_integral_number(),
+                          "layer shape input must have integral number element type, but has ",
```
Comment (tsocha) [inline src/ngraph/op/experimental/layers/prior_box.cpp]:
```suggestion
                          "layer shape input must be an integral number, but is: ",
```

File: src/ngraph/op/experimental/layers/prior_box.cpp [hunk 5]
```diff
@@ -37,14 +37,14 @@ void op::PriorBox::validate_and_infer_types()
...
                           layer_shape_et);
 
     auto image_shape_et = get_input_element_type(1);
     NODE_VALIDATION_CHECK(this,
-                          image_shape_et.compatible(element::Type_t::i64),
-                          "image shape input must have element type i64, but has ",
+                          image_shape_et.is_integral_number(),
+                          "image shape input must have integral number element type, but has ",
```
Comment (tsocha) [inline src/ngraph/op/experimental/layers/prior_box.cpp]:
```suggestion
                          "image shape input must be an integral number, but is: ",
```

File: src/ngraph/op/experimental/layers/prior_box_clustered.cpp [hunk 6]
```diff
@@ -37,14 +37,14 @@ void op::PriorBoxClustered::validate_and_infer_types()
     // shape node should have integer data type. For now we only allow i64
     auto layer_shape_et = get_input_element_type(0);
     NODE_VALIDATION_CHECK(this,
-                          layer_shape_et.compatible(element::Type_t::i64),
-                          "layer shape input must have element type i64, but has ",
+                          layer_shape_et.is_integral_number(),
+                          "layer shape input must have integral number element type, but has ",
```
Comment (tsocha) [inline src/ngraph/op/experimental/layers/prior_box_clustered.cpp]:
```suggestion
                          "layer shape input must be an integral number, but is: ",
```

File: src/ngraph/op/experimental/layers/prior_box_clustered.cpp [hunk 7]
```diff
@@ -37,14 +37,14 @@ void op::PriorBoxClustered::validate_and_infer_types()
...
                           layer_shape_et);
 
     auto image_shape_et = get_input_element_type(1);
     NODE_VALIDATION_CHECK(this,
-                          image_shape_et.compatible(element::Type_t::i64),
-                          "image shape input must have element type i64, but has ",
+                          image_shape_et.is_integral_number(),
+                          "image shape input must have integral number element type, but has ",
```
Comment (tsocha) [inline src/ngraph/op/experimental/layers/prior_box_clustered.cpp]:
```suggestion
                          "image shape input must be an integral number, but is: ",
```

File: src/ngraph/op/lrn.cpp [hunk 8]
```diff
@@ -105,8 +105,8 @@ void op::LRN::validate_and_infer_types()
 
     const auto& axes_type = get_input_element_type(1);
     NODE_VALIDATION_CHECK(this,
-                          axes_type.compatible(element::Type_t::i64),
-                          "Axes input must have element type i64 (axes type: ",
+                          axes_type.is_integral_number(),
+                          "Axes input must have integral number element type (axes type: ",
```
Comment (tsocha) [inline src/ngraph/op/lrn.cpp]:
```suggestion
                          "Axes input must be integral numbers, but are: ",
```

File: src/ngraph/op/pad.cpp [hunk 9]
```diff
@@ -252,14 +252,14 @@ void op::v1::Pad::validate_and_infer_types()
     }
 
     NODE_VALIDATION_CHECK(this,
-                          pads_begin_element_type.compatible(element::Type_t::i64),
-                          "pads_begin must be type i64 (axes type: ",
+                          pads_begin_element_type.is_integral_number(),
+                          "pads_begin must have integral number element type (axes type: ",
```
Comment (tsocha) [inline src/ngraph/op/pad.cpp]:
```suggestion
                          "pads_begin must be integral numbers, but are: ",
```

File: src/ngraph/op/pad.cpp [hunk 10]
```diff
@@ -252,14 +252,14 @@ void op::v1::Pad::validate_and_infer_types()
...
                           pads_begin_element_type,
                           ").");
 
     NODE_VALIDATION_CHECK(this,
-                          pads_end_element_type.compatible(element::Type_t::i64),
-                          "pads_end must be type i64 (axes type: ",
+                          pads_end_element_type.is_integral_number(),
+                          "pads_end must must have integral number element type (axes type: ",
```
Comment (tsocha) [inline src/ngraph/op/pad.cpp]:
```suggestion
                          "pads_end must be integral numbers, but are: ",
```

File: src/ngraph/op/reshape.cpp [hunk 11]
```diff
@@ -161,7 +161,7 @@ void op::v1::Reshape::validate_and_infer_types()
     auto pattern_et = get_input_element_type(1);
     // check data types
     NODE_VALIDATION_CHECK(
-        this, pattern_et.compatible(element::Type_t::i64), "Pattern must have element type i64.");
+        this, pattern_et.is_integral_number(), "Pattern must have integral number element type.");
```
Comment (tsocha) [inline src/ngraph/op/reshape.cpp]:
```suggestion
        this, pattern_et.is_integral_number(), "Pattern must be an integral number.");
```

Comment (tsocha) [reply src/ngraph/op/reshape.cpp]: Should we print a provided type as well as above?

File: src/ngraph/op/strided_slice.cpp [hunk 12]
```diff
@@ -71,12 +71,12 @@ void op::v1::StridedSlice::validate_and_infer_types()
     const auto& begin_mask_et = get_input_element_type(1);
     const auto& end_mask_et = get_input_element_type(2);
     NODE_VALIDATION_CHECK(this,
-                          begin_mask_et.compatible(element::Type_t::i64),
-                          "Begin mask must have element type i64, but has ",
+                          begin_mask_et.is_integral_number(),
+                          "Begin mask must have integral number element type, but has ",
```
Comment (tsocha) [inline src/ngraph/op/strided_slice.cpp]:
```suggestion
                          "Begin mask must be an integral number, but is: ",
```

File: src/ngraph/op/strided_slice.cpp [hunk 13]
```diff
@@ -71,12 +71,12 @@ void op::v1::StridedSlice::validate_and_infer_types()
...
                           begin_mask_et);
     NODE_VALIDATION_CHECK(this,
-                          end_mask_et.compatible(element::Type_t::i64),
-                          "End mask must have element type i64, but has ",
+                          end_mask_et.is_integral_number(),
+                          "End mask must have integral number element type, but has ",
```
Comment (tsocha) [inline src/ngraph/op/strided_slice.cpp]:
```suggestion
                          "End mask must be an integral number, but is: ",
```

File: test/type_prop/broadcast.cpp [hunk 14]
```diff
@@ -381,12 +381,12 @@ TEST(type_prop, broadcast_v1_broadcast_shape_et_wrong)
...
     {
         auto bc = make_shared<op::v1::Broadcast>(arg, bc_shape, bc_axes);
-        FAIL() << "Broadcast: did not detect shape element type not i64";
+        FAIL() << "Broadcast: did not detect shape element type not integral number";
     }
     catch (const NodeValidationFailure& error)
     {
         EXPECT_HAS_SUBSTRING(error.what(),
-                             std::string("Broadcast shape must have element type i64"));
+                             std::string("Broadcast shape must have integral number element type"));
```
Comment (tsocha) [inline test/type_prop/broadcast.cpp]:
```suggestion
                             std::string("Broadcast shape must be an integral number"));
```

File: test/type_prop/broadcast.cpp [hunk 15]
```diff
@@ -404,15 +404,15 @@ TEST(type_prop, broadcast_v1_axes_et_wrong)
...
     {
         auto bc = make_shared<op::v1::Broadcast>(arg, bc_shape, bc_axes);
-        FAIL() << "Broadcast: did not detect axes element type not i64";
+        FAIL() << "Broadcast: did not detect axes element type not integral number";
     }
     catch (const NodeValidationFailure& error)
     {
         EXPECT_HAS_SUBSTRING(error.what(),
-                             std::string("Broadcast axes must have element type i64"));
+                             std::string("Broadcast axes must have integral number element type"));
```
Comment (tsocha) [inline test/type_prop/broadcast.cpp]:
```suggestion
                             std::string("Broadcast axes must be an integral number"));
```

File: test/type_prop/broadcast.cpp [hunk 16]
```diff
@@ -404,15 +404,15 @@ TEST(type_prop, broadcast_v1_axes_et_wrong)
...
     }
     catch (...)
     {
         FAIL() << "Deduced type check failed for unexpected reason";
     }
-}
+}
```
Comment (tsocha) [inline test/type_prop/broadcast.cpp]:
```suggestion
}

```

File: test/type_prop/pad.cpp [hunk 17]
```diff
@@ -571,7 +571,9 @@ TEST(type_prop, pad_v1_arg_pads_begin_incompatible_type)
     }
     catch (const NodeValidationFailure& error)
     {
-        EXPECT_HAS_SUBSTRING(error.what(), std::string("pads_begin must be type i64 (axes type:"));
+        EXPECT_HAS_SUBSTRING(
+            error.what(),
+            std::string("pads_begin must have integral number element type (axes type:"));
```
Comment (tsocha) [inline test/type_prop/pad.cpp]:
```suggestion
            std::string("pads_begin must be integral numbers, but are:"));
```

File: test/type_prop/pad.cpp [hunk 18]
```diff
@@ -594,7 +596,9 @@ TEST(type_prop, pad_v1_arg_pads_end_incompatible_type)
     }
     catch (const NodeValidationFailure& error)
     {
-        EXPECT_HAS_SUBSTRING(error.what(), std::string("pads_end must be type i64 (axes type:"));
+        EXPECT_HAS_SUBSTRING(
+            error.what(),
+            std::string("pads_end must must have integral number element type (axes type:"));
```
Comment (tsocha) [inline test/type_prop/pad.cpp]:
```suggestion
            std::string("pads_end must an integral numbers, but are:"));
```

File: test/type_prop/strided_slice.cpp [hunk 19]
```diff
@@ -37,7 +37,8 @@ TEST(type_prop, strided_slice_begin_incorrect_type)
     }
     catch (const NodeValidationFailure& error)
     {
-        EXPECT_HAS_SUBSTRING(error.what(), std::string("Begin mask must have element type i64"));
+        EXPECT_HAS_SUBSTRING(error.what(),
+                             std::string("Begin mask must have integral number element type"));
```
Comment (tsocha) [inline test/type_prop/strided_slice.cpp]:
```suggestion
                             std::string("Begin mask must be an integral number"));
```

File: test/type_prop/strided_slice.cpp [hunk 20]
```diff
@@ -59,7 +60,8 @@ TEST(type_prop, strided_slice_end_incorrect_type)
     }
     catch (const NodeValidationFailure& error)
     {
-        EXPECT_HAS_SUBSTRING(error.what(), std::string("End mask must have element type i64"));
+        EXPECT_HAS_SUBSTRING(error.what(),
+                             std::string("End mask must have integral number element type"));
```
Comment (tsocha) [inline test/type_prop/strided_slice.cpp]:
```suggestion
                             std::string("End mask must be an integral number"));
```

**Pull request merged**